va/va.go

```diff
@@ -670,10 +673,16 @@ func (va *ValidationAuthorityImpl) PerformValidation(ctx context.Context, req *v
 				va.metrics.remoteValidationFailures.Inc()
 			} else {
 				challenge.Status = core.StatusValid
+				// Timestamp the valid challenge.
+				currentTime := time.Now().UTC()
+				challenge.Validated = &currentTime
 			}
 		}
 	} else {
 		challenge.Status = core.StatusValid
+		// Timestamp the valid challenge.
+		currentTime := time.Now().UTC()
+		challenge.Validated = &currentTime
```

Comment:
The face that this comment and the corresponding code exists in three places means that if it ever needs to be edited or updated the author will need to address multiple lines. I believe that these three lines alone could each be moved to a helper function.

Additionally, the logic here has grown pretty large and could be difficult to reason about. It might make sense to move the code for each of these conditions to their own separate testable function.

Reply:
Also, I think we should record the validated timestamp as the time we _began_ validation rather than the time we finished it. I think that should simplify this code a bit on its own.

Reply:
> Also, I think we should record the validated timestamp as the time we _began_ validation rather than the time we finished it. I think that should simplify this code a bit on its own.

I'm not sure that the RFC language supports recording the validation timestamp at the time that we began validation:
```
https://tools.ietf.org/html/rfc8555#section-8:
...
validated (optional, string):  The time at which the server validated
    this challenge, encoded in the format specified in [RFC3339].
    This field is REQUIRED if the "status" field is "valid".
```

Reply:
> Also, I think we should record the validated timestamp as the time we _began_ validation rather than the time we finished it. I think that should simplify this code a bit on its own.

Wouldn't that end with us having validations that ended in statusInvalid but have a validated timestamp?

Reply:
> I'm not sure that the RFC language supports recording the validation timestamp at the time that we began validation:

There's no way to interpret "validation" as a single point in time; it's a process that occurs between t and time t'. If we had wanted to very precisely express that exact time range in the spec, we could perhaps have written "validationBegan" and "validationEnded." But even that is extremely subjective. We might think we began when we sent out the first DNS query packet to Unbound, but the outside world doesn't see anything to milliseconds later.

Given there's no one perfect time, I think we are fine to pick anything roughly between t and t'. As this field is mainly a debugging aid, I think it's most useful for this time to be pretty close to the time ACME clients will have recorded in their log that they POSTed to the challenge.

> Wouldn't that end with us having validations that ended in statusInvalid but have a validated timestamp?

You could store the current time when we begin validation, and only assign it to the validationRecord upon success.

Also, BTW: we talked about only storing (in the DB) the validated timestamp on success, since that's the only time we want to send it to the user, and we want to economize on DB space. However, for the log, it's nice to log the beginning of the validation process regardless. That way when we do logs analysis we can plot (e.g.) failure count by time validation began, which is much more meaningful than plotting it by the time the validation ended. We already have the time the validation ended - that's in the log timestamp. We also have the validation latency, and could back-calculate when the validation must have begun, but it's nice to just have the value straight out.

Reply:
I changed to using the existing vStart clock that we take earlier in the function. Samantha and I worked out a helper function that sets a challenge to valid and a timestamp which seemed nice when creating a new time.Now(), but after refactoring to use vStart it seemed like more complexity to pass in time and challenge pointers just to do add one more line of code to set the validated time. With the new function there ended up being twice as many lines added, as well as a function that would likely need tests. The latest commit just removes the currentTime lines and passes in the vStart time to the validated field.

If there is a strong preference for the helper function I can do that.